In [1]:
!pip install tabletext
import tabletext
import pybgpstream
import pprint
import numpy as np
from operator import itemgetter
from itertools import groupby

Primero averiguamos el ASN (Autonomous System Number) de nuestra ISP:

Entrando a: https://bgp.he.net 

<img src="imagenes/ASN.png"> 

**AS objetivo**

Aplicando el cálculo dado a nuestros números de padrón (95457 y 96626) obtenemos la clave 9, por lo que para el presente trabajo práctico utilizaremos el sistema autónomo de la empresa de servicios en la nube neozelandesa Revera.
**ASN = 18353**

In [2]:
REVERA_ASN=18353
ISP_ASN=10318

Colectores a utilizar: **route-views2** y **route-views.saopaulo** 


In [3]:
COLLECTOR_1='route-views2'
COLLECTOR_2='route-views.saopaulo'

In [4]:
# ’’’
# export LD_LIBRARY_PATH =/usr/local/lib
# ’’’

#---------------------------------------------------------
# Definiciones

TIME_INIT="2017-03-01"
TIME_END="2017-03-01 00:05"

In [5]:
def getAsPaths(collector, asn):
    #---------------------------------------------------------
    # Genero la consulta a la API
    stream = pybgpstream.BGPStream (
    from_time = TIME_INIT ,
    until_time = TIME_END ,
    filter = "type  ribs  and  collector  %s  and  path  %s" % (collector,asn))
    
    # Itero en cada uno de los anuncios
    as_paths = []
    elements = []
    for elem in stream :
        
    # Filtro solo los AS-PATHS que tengan como destino
    # (ultimo AS en la cadena) a Telecom Argentina S.A
        as_path = elem.fields["as-path"].split(' ')
        if int (as_path[ -1]) == asn :
            
            #Eliminamos los ASN consecutivos repetidos
            as_path_without_consecutives_duplicated = tuple(map(itemgetter(0), groupby(as_path)))
            as_paths.append(as_path_without_consecutives_duplicated)
            elements.append(elem)
            #pprint.pprint(elem.fields)
    return as_paths, elements

In [6]:
as_paths_1, elem_1 = getAsPaths(COLLECTOR_1, ISP_ASN)
as_paths_1_unique = set(as_paths_1)

In [7]:
as_paths_2, elem_2 = getAsPaths(COLLECTOR_2, ISP_ASN)
as_paths_2_unique = set(as_paths_2)

In [8]:
as_paths_3, elem_3 = getAsPaths(COLLECTOR_1, REVERA_ASN)
as_paths_3_unique = set(as_paths_3)

In [9]:
as_paths_4, elem_4 = getAsPaths(COLLECTOR_2, REVERA_ASN)
as_paths_4_unique = set(as_paths_4)

¿Cuántos AS-PATHS diferentes se ven desde cada colector hacia el AS objetivo y hacia su ISP? Se considera un AS-PATH a toda la cadena de ASes. Usted debe mostrar 4 resultados (2 colectores, 2 ASes)

* Colector: **route-views2**
* AS: **ISP**
* AS - PATHS: 

In [24]:
print(tabletext.to_text(as_paths_1_unique))

┌────────┬───────┬───────┬───────┬───────┬───────┐
│ 5413   │ 3356  │ 10481 │ 10318 │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┤
│ 7660   │ 2516  │ 3549  │ 10481 │ 10318 │       │
├────────┼───────┼───────┼───────┼───────┼───────┤
│ 852    │ 3356  │ 6762  │ 10481 │ 10318 │       │
├────────┼───────┼───────┼───────┼───────┼───────┤
│ 1403   │ 3257  │ 6762  │ 10481 │ 10318 │       │
├────────┼───────┼───────┼───────┼───────┼───────┤
│ 31019  │ 39326 │ 1299  │ 6762  │ 10481 │ 10318 │
├────────┼───────┼───────┼───────┼───────┼───────┤
│ 3130   │ 1239  │ 3356  │ 10481 │ 10318 │       │
├────────┼───────┼───────┼───────┼───────┼───────┤
│ 20771  │ 47872 │ 2914  │ 6762  │ 10481 │ 10318 │
├────────┼───────┼───────┼───────┼───────┼───────┤
│ 18106  │ 6453  │ 6762  │ 10481 │ 10318 │       │
├────────┼───────┼───────┼───────┼───────┼───────┤
│ 53364  │ 3257  │ 3356  │ 10481 │ 10318 │       │
├────────┼───────┼───────┼───────┼───────┼───────┤
│ 5413   │ 1299  │ 6762  │ 1048

In [36]:
len(as_paths_1_unique)

100

* Colector: **route-views.saopaulo**
* AS: **ISP**
* AS - PATHS: 


In [25]:
print(tabletext.to_text(as_paths_2_unique))

┌────────┬───────┬───────┬───────┬───────┬───────┬───────┐
│ 53013  │ 4809  │ 1299  │ 6762  │ 10481 │ 10318 │       │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 28329  │ 6762  │ 10481 │ 10318 │       │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 52940  │ 53171 │ 3549  │ 10481 │ 10318 │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 1916   │ 174   │ 6762  │ 10481 │ 10318 │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 262750 │ 28329 │ 3549  │ 10481 │ 10318 │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 52863  │ 6762  │ 3549  │ 10481 │ 10318 │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 264268 │ 28329 │ 6762  │ 10481 │ 10318 │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 28138  │ 16735 │ 3549  │ 10481 │ 10318 │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┼──────

In [37]:
len(as_paths_2_unique)

42

* Colector: **route-views2**
* AS: **REVERA**
* AS - PATHS: 

In [26]:
print(tabletext.to_text(as_paths_3_unique))

┌────────┬───────┬───────┬───────┬───────┬───────┐
│ 286    │ 2914  │ 4648  │ 18353 │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┤
│ 852    │ 9500  │ 9901  │ 18353 │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┤
│ 47872  │ 6939  │ 9500  │ 9901  │ 18353 │       │
├────────┼───────┼───────┼───────┼───────┼───────┤
│ 11686  │ 9500  │ 9901  │ 18353 │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┤
│ 3741   │ 6939  │ 9500  │ 9901  │ 18353 │       │
├────────┼───────┼───────┼───────┼───────┼───────┤
│ 200130 │ 1299  │ 4648  │ 18353 │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┤
│ 7018   │ 2914  │ 4648  │ 18353 │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┤
│ 3277   │ 3267  │ 6939  │ 9500  │ 9901  │ 18353 │
├────────┼───────┼───────┼───────┼───────┼───────┤
│ 13030  │ 1273  │ 9500  │ 9901  │ 18353 │       │
├────────┼───────┼───────┼───────┼───────┼───────┤
│ 3549   │ 3356  │ 2914  │ 4648

* Colector: **route-views.saopaulo**
* AS: **REVERA**
* AS - PATHS: 

In [27]:
print(tabletext.to_text(as_paths_4_unique))

┌────────┬────────┬───────┬───────┬───────┬───────┬───────┐
│ 52720  │ 6939   │ 9500  │ 9901  │ 18353 │       │       │
├────────┼────────┼───────┼───────┼───────┼───────┼───────┤
│ 14840  │ 6939   │ 9500  │ 9901  │ 18353 │       │       │
├────────┼────────┼───────┼───────┼───────┼───────┼───────┤
│ 1916   │ 6939   │ 9500  │ 9901  │ 18353 │       │       │
├────────┼────────┼───────┼───────┼───────┼───────┼───────┤
│ 52888  │ 1916   │ 2914  │ 4648  │ 18353 │       │       │
├────────┼────────┼───────┼───────┼───────┼───────┼───────┤
│ 262757 │ 6939   │ 4648  │ 18353 │       │       │       │
├────────┼────────┼───────┼───────┼───────┼───────┼───────┤
│ 264268 │ 28329  │ 6762  │ 2914  │ 4648  │ 18353 │       │
├────────┼────────┼───────┼───────┼───────┼───────┼───────┤
│ 52940  │ 53171  │ 3549  │ 3356  │ 1299  │ 4648  │ 18353 │
├────────┼────────┼───────┼───────┼───────┼───────┼───────┤
│ 264268 │ 6939   │ 9500  │ 9901  │ 18353 │       │       │
├────────┼────────┼───────┼───────┼─────

¿Cuáles son los proveedores observados para el AS objetivo y para su ISP? Se considera proveedor al AS inmediatamente anterior en la cadena de AS-PATH. En la medida de lo posible trate de decir el ASname de los proveedores. 

In [28]:
ISP_providers = []
for as_path in list(as_paths_1_unique) + list(as_paths_2_unique):
    ISP_providers.append(as_path[-2])


In [29]:
set(ISP_providers)

{'10481'}

In [30]:
revera_providers = []
for as_path in list(as_paths_3_unique) + list(as_paths_4_unique):
    revera_providers.append(as_path[-2])

In [31]:
set(revera_providers)


{'4648', '9901'}

En https://bgp.potaroo.net/cidr/autnums.html los AS names

	AS: ISP
	Proveedor: 
        AS Number: 10481
        AS Name: Prima S.A., AR

	AS: Revera
	Proveedores: 
		AS Number:4648
		AS Name: SPARK-NZ Global-Gateway Internet, NZ

		AS Number:9901
		AS Name: VFNZ-TRANSIT-AS Vodafone NZ Ltd, NZ


¿Cuántos prefijos diferentes anuncia el AS objetivo y cuantos su ISP?

In [32]:
prefixes_ISP = [] 
for elem in elem_1 + elem_2:
    prefixes_ISP.append(elem.fields['prefix'])
set(prefixes_ISP)

{'152.168.0.0/19',
 '152.168.128.0/19',
 '152.168.160.0/19',
 '152.168.192.0/19',
 '152.168.224.0/19',
 '152.168.32.0/19',
 '152.168.64.0/19',
 '152.168.96.0/19',
 '152.169.0.0/19',
 '152.169.128.0/19',
 '152.169.160.0/19',
 '152.169.192.0/19',
 '152.169.224.0/19',
 '152.169.32.0/19',
 '152.169.64.0/19',
 '152.169.96.0/19',
 '152.170.0.0/19',
 '152.170.128.0/19',
 '152.170.160.0/19',
 '152.170.192.0/19',
 '152.170.224.0/19',
 '152.170.32.0/19',
 '152.170.64.0/19',
 '152.170.96.0/19',
 '152.171.0.0/19',
 '152.171.128.0/19',
 '152.171.160.0/19',
 '152.171.192.0/19',
 '152.171.224.0/19',
 '152.171.32.0/19',
 '152.171.64.0/19',
 '152.171.96.0/19',
 '170.231.136.0/22',
 '170.231.136.0/23',
 '170.231.138.0/23',
 '181.164.0.0/14',
 '181.164.0.0/19',
 '181.164.128.0/19',
 '181.164.160.0/19',
 '181.164.192.0/19',
 '181.164.224.0/19',
 '181.164.32.0/19',
 '181.164.64.0/19',
 '181.164.96.0/19',
 '181.165.0.0/19',
 '181.165.128.0/19',
 '181.165.160.0/19',
 '181.165.192.0/19',
 '181.165.224.0/19',


In [33]:
prefixes_revera = [] 
for elem in elem_3 + elem_4:
    prefixes_revera.append(elem.fields['prefix'])
set(prefixes_revera)

{'103.253.48.0/24',
 '103.253.50.0/24',
 '202.14.96.0/24',
 '202.36.252.0/24',
 '202.36.253.0/24',
 '202.36.254.0/24',
 '202.36.30.0/23',
 '202.37.205.0/24',
 '202.37.64.0/24',
 '202.37.65.0/24',
 '202.55.100.0/24',
 '202.55.101.0/24',
 '202.55.102.0/24',
 '202.55.103.0/24',
 '202.55.104.0/24',
 '202.55.105.0/24',
 '202.55.106.0/24',
 '202.55.107.0/24',
 '202.55.108.0/24',
 '202.55.109.0/24',
 '202.55.111.0/24',
 '202.55.96.0/24',
 '202.55.97.0/24',
 '202.55.98.0/24',
 '202.55.99.0/24',
 '220.247.132.0/24',
 '220.247.133.0/24',
 '2404:e200:101::/48',
 '2404:e200:102::/48',
 '2404:e200:103::/48',
 '2404:e200:104::/48',
 '2404:e200:1::/48',
 '2404:e200:6::/48'}

Repita todos los puntos anteriores pero para el año 2012. En caso de haber diferencias, detállelas.

In [38]:
TIME_INIT="2012-03-01"
TIME_END="2012-03-01 00:05"

In [39]:
as_paths_1, elem_1 = getAsPaths(COLLECTOR_1, ISP_ASN)
as_paths_1_unique = set(as_paths_1)

In [40]:
as_paths_2, elem_2 = getAsPaths(COLLECTOR_2, ISP_ASN)
as_paths_2_unique = set(as_paths_2)

In [41]:
as_paths_3, elem_3 = getAsPaths(COLLECTOR_1, REVERA_ASN)
as_paths_3_unique = set(as_paths_3)

In [42]:
as_paths_4, elem_4 = getAsPaths(COLLECTOR_2, REVERA_ASN)
as_paths_4_unique = set(as_paths_4)

¿Cuántos AS-PATHS diferentes se ven desde cada colector hacia el AS objetivo y hacia su ISP? Se considera un AS-PATH a toda la cadena de ASes. Usted debe mostrar 4 resultados (2 colectores, 2 ASes)

* Colector: **route-views2**
* AS: **ISP**
* AS - PATHS: 

In [43]:
print(tabletext.to_text(as_paths_1_unique))

┌───────┬───────┬───────┬───────┬───────┬───────┬───────┬───────┐
│ 852   │ 3561  │ 12956 │ 22927 │ 10481 │ 10318 │       │       │
├───────┼───────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 13030 │ 3549  │ 10481 │ 10318 │       │       │       │       │
├───────┼───────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 3303  │ 1299  │ 3549  │ 10481 │ 10318 │       │       │       │
├───────┼───────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 3257  │ 3549  │ 10481 │ 10318 │       │       │       │       │
├───────┼───────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 6539  │ 577   │ 6453  │ 3549  │ 10481 │ 10318 │       │       │
├───────┼───────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 7660  │ 2516  │ 3549  │ 10481 │ 10318 │       │       │       │
├───────┼───────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 8001  │ 4436  │ 701   │ 12956 │ 22927 │ 10481 │ 10318 │       │
├───────┼───────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 5413  │ 

* Colector: **route-views.saopaulo**
* AS: **ISP**
* AS - PATHS: 


In [44]:
print(tabletext.to_text(as_paths_2_unique))

┌────────┬───────┬───────┬───────┬───────┬───────┬───────┐
│ 16735  │ 3549  │ 10481 │ 10318 │       │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 28289  │ 53131 │ 10429 │ 12956 │ 22927 │ 10481 │ 10318 │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 262757 │ 16735 │ 12956 │ 22927 │ 10481 │ 10318 │       │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 262757 │ 16735 │ 3549  │ 10481 │ 10318 │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 28138  │ 16735 │ 12956 │ 22927 │ 10481 │ 10318 │       │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 28289  │ 53131 │ 16735 │ 3549  │ 10481 │ 10318 │       │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 16735  │ 12956 │ 22927 │ 10481 │ 10318 │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 28138  │ 16735 │ 3549  │ 10481 │ 10318 │       │       │
└────────┴───────┴───────┴───────┴───────┴───────┴──────

* Colector: **route-views2**
* AS: **REVERA**
* AS - PATHS: 

In [45]:
print(tabletext.to_text(as_paths_3_unique))

┌───────┬───────┬───────┬───────┬───────┬───────┐
│ 3549  │ 4637  │ 9901  │ 18353 │       │       │
├───────┼───────┼───────┼───────┼───────┼───────┤
│ 6539  │ 4648  │ 18353 │       │       │       │
├───────┼───────┼───────┼───────┼───────┼───────┤
│ 7018  │ 2914  │ 4648  │ 18353 │       │       │
├───────┼───────┼───────┼───────┼───────┼───────┤
│ 3549  │ 701   │ 703   │ 9901  │ 18353 │       │
├───────┼───────┼───────┼───────┼───────┼───────┤
│ 701   │ 3549  │ 4826  │ 4770  │ 18353 │       │
├───────┼───────┼───────┼───────┼───────┼───────┤
│ 2914  │ 701   │ 703   │ 9901  │ 18353 │       │
├───────┼───────┼───────┼───────┼───────┼───────┤
│ 5413  │ 6939  │ 4826  │ 4770  │ 18353 │       │
├───────┼───────┼───────┼───────┼───────┼───────┤
│ 1221  │ 4637  │ 9901  │ 18353 │       │       │
├───────┼───────┼───────┼───────┼───────┼───────┤
│ 2914  │ 4648  │ 18353 │       │       │       │
├───────┼───────┼───────┼───────┼───────┼───────┤
│ 2497  │ 4637  │ 9901  │ 18353 │       │       │


* Colector: **route-views.saopaulo**
* AS: **REVERA**
* AS - PATHS: 

In [46]:
print(tabletext.to_text(as_paths_4_unique))

┌────────┬───────┬───────┬───────┬───────┬───────┬───────┬───────┐
│ 262757 │ 16735 │ 12989 │ 4648  │ 18353 │       │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 16735  │ 12989 │ 4648  │ 18353 │       │       │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 16735  │ 6939  │ 4826  │ 4770  │ 18353 │       │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 28289  │ 53131 │ 10429 │ 12956 │ 701   │ 703   │ 9901  │ 18353 │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 28138  │ 16735 │ 6939  │ 4826  │ 4770  │ 18353 │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 28289  │ 53131 │ 16735 │ 1239  │ 4648  │ 18353 │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┼───────┤
│ 262757 │ 16735 │ 1239  │ 4648  │ 18353 │       │       │       │
├────────┼───────┼───────┼───────┼───────┼───────┼───────┼────

¿Cuáles son los proveedores observados para el AS objetivo y para su ISP? Se considera proveedor al AS inmediatamente anterior en la cadena de AS-PATH. En la medida de lo posible trate de decir el ASname de los proveedores. 

In [47]:
ISP_providers = []
for as_path in list(as_paths_1_unique) + list(as_paths_2_unique):
    ISP_providers.append(as_path[-2])


In [48]:
set(ISP_providers)

{'10481'}

In [49]:
revera_providers = []
for as_path in list(as_paths_3_unique) + list(as_paths_4_unique):
    revera_providers.append(as_path[-2])

In [50]:
set(revera_providers)

{'4648', '4770', '9901'}

En https://bgp.potaroo.net/cidr/autnums.html los AS names

	AS: ISP
	Proveedor: 
        AS Number: 10481
        AS Name: Prima S.A., AR

	AS: Revera
	Proveedores: 
		AS Number:4648
		AS Name: SPARK-NZ Global-Gateway Internet, NZ

		AS Number:4770
		AS Name: ICONZ-AS ICONZ Ltd, NZ
        
        AS Number:9901
		AS Name: VFNZ-TRANSIT-AS Vodafone NZ Ltd, NZ


¿Cuántos prefijos diferentes anuncia el AS objetivo y cuantos su ISP?

In [51]:
prefixes_ISP = [] 
for elem in elem_1 + elem_2:
    prefixes_ISP.append(elem.fields['prefix'])
set(prefixes_ISP)

{'181.164.0.0/14',
 '181.164.0.0/15',
 '181.164.0.0/19',
 '181.164.128.0/19',
 '181.164.160.0/19',
 '181.164.192.0/19',
 '181.164.224.0/19',
 '181.164.32.0/19',
 '181.164.64.0/19',
 '181.164.96.0/19',
 '181.28.0.0/14',
 '186.136.0.0/14',
 '186.136.0.0/15',
 '186.136.0.0/19',
 '186.136.128.0/19',
 '186.136.160.0/19',
 '186.136.192.0/19',
 '186.136.224.0/19',
 '186.136.32.0/19',
 '186.136.64.0/19',
 '186.136.96.0/19',
 '186.137.0.0/19',
 '186.137.128.0/19',
 '186.137.160.0/19',
 '186.137.192.0/19',
 '186.137.224.0/19',
 '186.137.32.0/19',
 '186.137.64.0/19',
 '186.137.96.0/19',
 '190.16.0.0/14',
 '190.16.0.0/19',
 '190.16.128.0/19',
 '190.16.144.0/20',
 '190.16.160.0/19',
 '190.16.192.0/19',
 '190.16.224.0/19',
 '190.16.32.0/19',
 '190.16.64.0/19',
 '190.16.96.0/19',
 '190.17.0.0/19',
 '190.17.128.0/19',
 '190.17.160.0/19',
 '190.17.192.0/19',
 '190.17.224.0/19',
 '190.17.32.0/19',
 '190.17.64.0/19',
 '190.17.96.0/19',
 '190.18.0.0/19',
 '190.18.128.0/19',
 '190.18.160.0/19',
 '190.18.19

In [52]:
prefixes_revera = [] 
for elem in elem_3 + elem_4:
    prefixes_revera.append(elem.fields['prefix'])
set(prefixes_revera)

{'192.88.85.0/24',
 '202.14.96.0/24',
 '202.36.30.0/23',
 '202.37.205.0/24',
 '202.55.100.0/24',
 '202.55.101.0/24',
 '202.55.102.0/24',
 '202.55.104.0/24',
 '202.55.105.0/24',
 '202.55.106.0/24',
 '202.55.107.0/24',
 '202.55.96.0/24',
 '202.55.97.0/24',
 '202.55.98.0/24',
 '202.55.99.0/24'}